### ISSUE TITLE
Is it feasible to control the focus in talkback mode via 'adb tap'?

### ISSUE THREAD

**Issue by PeiLeizzz** [qbalsdon/accessibility_broadcast_dev#1]
According to my experiments, the `adb tap` cannot be received by the accessibility layer, and it will go directly into the box instead of focusing on it.
How should I do?

**Comment by qbalsdon** [comment]
Hey there!

Thanks for noticing this issue - the Service was not getting ViewId's because I had refactored the service config XML. It needed to be changed from 

```
android:accessibilityFlags="flagDefault"
```
to
```
android:accessibilityFlags="flagDefault|flagReportViewIds"
```

Which is odd because I suspected `android:canRetrieveWindowContent="true"` to be able to get the id's, but alas no.

So you can now do this in a few ways:
```
adb shell am broadcast -a com.balsdon.talkback.accessibility -e ACTION "ACTION_FOCUS_ELEMENT" [OPTION]

// OPTION 1:
    -e PARAMETER_ID [VIEW_ID_NAME] // by id, without package name - e.g. "openMenu"

// OPTION 2:
    -e PARAMETER_TEXT [DISPLAYED_TEXT] // by the text attribute, e.g. "curtain"
 
// OPTION 3:
    -e PARAMETER_TYPE "element type" --e DIRECTION "[DIRECTION_FORWARD | DIRECTION_BACK]" // searches nodes by class name 
 
// OPTION 4:
    -e PARAMETER_HEADING "[DIRECTION_FORWARD | DIRECTION_BACK]" // just cycle through headings
  
// FULL EXAMPLE
    adb shell am broadcast -a com.balsdon.talkback.accessibility -e ACTION "ACTION_FOCUS_ELEMENT" -e PARAMETER_ID openMenu 
```

**PLEASE NOTE:** This will NOT scroll the screen if the node is off screen - it's a known issue in the TODOs.

I hope the latest commit resolves your issue!

Additionally, I am moving away from this solution - I don't like the gesture / action disconnect and the variation of TalkBack versions that exist - Pixel (supported), Samsung (supported), Amazon (not supported). I have moved on to using a [custom fork of TalkBack][0] to resolve the disconnect and will be porting the curtain there as well over time. Then this project will be deprecated. 

[0]: https://github.com/qbalsdon/talkback

**Comment by qbalsdon** [comment]
If you are wondering why `adb shell input tap [x,y]` doesn't work for accessibility focus, I wrote a [blog explaining the accessibility layer][0], and that adb bypasses this layer.

[0]: https://qbalsdon.github.io/android/accessibility/talkback/2022/07/06/android_talkback.html

**Comment by PeiLeizzz** [comment]
> If you are wondering why `adb shell input tap [x,y]` doesn't work for accessibility focus, I wrote a [blog explaining the accessibility layer](https://qbalsdon.github.io/android/accessibility/talkback/2022/07/06/android_talkback.html), and that adb bypasses this layer.

Thank you very much for your answer! 
The field of accessibility is a very niche but actually important one, and I searched the web for a lot of explanations about this problem, but in the end, it failed. 
Fortunately, your article gave me a good introduction to the internal mechanism of accessibility technology. Regarding the method you gave, I will try! Thank you! 

And, my current job is to automatically detect the accessibility of apps on the market, here I can recommend a software: `Across Center`, which can control Android and iOS in the form of sharing the computer's keyboard and mouse.
Through this software, you can switch the focus forward or back or many other command(e.g. return to the previous page / return to the first focus box ...) by sending keyboard shortcuts through code, but unfortunately, it seems that it is impossible to send mouse click with code to switch focus to a specified location `[x, y]`(because it will not be received by the accessibility layer), so I raised this issue.

**Comment by qbalsdon** [comment]
Yeah I like to use [Genymotion's scrcpy][0] to do my device screen mirroring. I see that [Across Center][1] is similar, except it's like remoting in from another network. I think you could also use [TeamViewer][2]

I'm going to close this issue, but feel free to message and chat. My job is Android Accessibility and I focus on making tools for developers and testers to make it easier for them. A lot of the work I produce is open source, like this one.

[0]: https://github.com/Genymobile/scrcpy
[1]: http://www.acrosscenter.com/
[2]: https://www.teamviewer.com/en/solutions/support-mobile-devices/
